### PR TITLE
[expo-wallet] Supports adding passes to Apple Wallet unimodule

### PR DESCRIPTION
**Pull request by vivianzzhu91**
# Why

Due to this [feature request](https://expo.canny.io/feature-requests/p/apple-wallet), we decided to add `expo-wallet` module to add passes to Apple Wallet with different functionalities and UI component (the Add to Apple Wallet button) and web implementations.

# How

With reference to Apple official [docs](https://developer.apple.com/documentation/passkit/wallet?language=objc) and `react-native-wallet`, I created this module.

See also https://github.com/expo/expo-cli/pull/942.

# Test Plan

Tested on real devices and got all expected behaviors and event emitters working.
IMG_6852

### CODE REVIEW

**Comment by hesyifei** [comment]
Todo:

- [ ] Add [PKAddPassButton](https://developer.apple.com/documentation/passkit/pkaddpassbutton)

- [ ] Web support (`Wallet.addPassFromUrlAsync(url)` should open the URL. Probably add a `alert` if the user is not using Safari to say that they have to open the downloaded .pkpass file manually.)

**Comment by sheepsteak** [comment]
This is great! Will this make it into [Expo SDK 35?](https://github.com/expo/expo/issues/5564)

**Comment by esamelson** [comment]
@sheepsteak - no, it will not; we have a pretty hard deadline for SDK 35 due to iOS 13, and this PR has not gone through review yet. However, this is going to be a bare-workflow-only module for the time being (just like `expo-in-app-payments`) and so at the time it does land, we will not need to wait for the next SDK version to release it.

**Comment by dbbk** [comment]
Any news on this? This would be very useful for us.

**Comment by amsimoes** [comment]
Any news? I need this very urgently on my project, otherwise I'll have to eject :'(

**Comment by AngSin** [comment]
Can we get an update about this please?

**Comment by esamelson** [comment]
Hi all - we are not actively working on this right now. You are welcome to check out this branch and try installing the module yourself but ymmv.

File: packages/expo-wallet/android/src/main/java/expo/modules/wallet/WalletViewManager.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  @Override
+  public ViewManagerType getViewManagerType() {
+    return ViewManagerType.SIMPLE;
+  }
+
+  @Override
+  public List<String> getExportedEventNames() {
+    return Arrays.asList("onSomethingHappened");
```
Comment (EvanBacon) [inline packages/expo-wallet/android/src/main/java/expo/modules/wallet/WalletViewManager.java]: ...

File: packages/expo-wallet/package.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  },
+  "author": "650 Industries, Inc.",
+  "license": "MIT",
+  "homepage": "https://github.com/expo/expo/tree/master/packages/expo-wallet",
+  "peerDependencies": {
+    "@unimodules/core": "~1.0.0"
+  },
+  "gitHead": "9518929d6f2ba9cf8e0aae81d34cc1eb52f7093a"
+
+,"_resolved": ""
```
Comment (EvanBacon) [inline packages/expo-wallet/package.json]: remove

File: packages/expo-wallet/src/ExpoWallet.web.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+export default {
+  async canAddPassesAsync(): Promise<boolean> {
+    return true;
+  },
+  async addPassFromUrlAsync(url: string): Promise<boolean> {
+    window.open(url);
```
Comment (EvanBacon) [inline packages/expo-wallet/src/ExpoWallet.web.ts]: nice

**Review by EvanBacon — CHANGES_REQUESTED** [review]
Android shims need to be resolved

**Comment by amsimoes** [comment]
Guys, any news? @vivianzzhu91 @EvanBacon how can I help here?

**Comment by sjchmiela** [comment]
In an effort to keep all the Expo libraries high quality and well maintained we sometimes have to trim the branches of different libraries we would like to support, but don't really have the bandwidth to.

Expo bare workflow and `react-native-wallet` will be the best way to go if you need this feature. 🙂

**Comment by HughBerriez** [comment]
@sjchmiela  This feature is the one feature in the next 3 months or so that I forsee causing us to have to eject to the bare workflow. It looks like so much work was done on the PR for the adding card to wallets feature. 

Is this really closed for the foreseeable future? Definitely a bit of a bummer if so

**Comment by sjchmiela** [comment]
It does indeed look like a good work, unfortunately I'm afraid we wouldn't feel comfortable taking on yet another module's maintaining burden. Once we tidy up what we have already we may revisit this PR.

FWIW even if this was implemented there's no 100% chance that this would be available in the managed workflow — I haven't checked, but Wallet stuff seems like it may cause apps to undergo a more thorough Apple Review process and not all Expo-built apps would be able to explain why do they need to do something with the Apple Wallet API.

Moreover, bare workflow has undergone some pretty serious changes since the middle of 2019 and ejecting to bare workflow no longer means "having to spend a week configuring some strange native stuff" nor "not being able to use all the Expo goodness" — we've removed ExpoKit so now you eject to bare-bare-React-Native with unimodules preconfigured and we've moved last of our best features to bare-compatible packages, i.e. `expo-updates`, `expo-splash-screen` and `expo-notifications`.

**Comment by HughBerriez** [comment]
@sjchmiela thanks for the quick and detailed reply. That makes a lot of sense - I'll take a deeper look into the updated bare workflow - sounds much less daunting that a few years ago. Really do appreciate the insights above - expo is incredible.

**Pull request closed without merging**